File: go/rules.go
package main

import (
	"fmt"
	"log"
	"os"
	"github.com/algolia/algoliasearch-client-go/v4/algolia/search"
	"github.com/joho/godotenv"
)

func PrintErrAndExit(err error) {
	fmt.Println(err)
	os.Exit(1)
}

func main() {
	err := godotenv.Load()
	if err != nil {
		log.Fatal("Error loading .env file")
	}
	// Algolia client credentials
	appID, apiKey, indexName := os.Getenv("ALGOLIA_APP_ID"), os.Getenv("ALGOLIA_API_KEY"), os.Getenv("ALGOLIA_INDEX_NAME")

	// Start the API client
	// https://www.algolia.com/doc/libraries/sdk/methods/search#go
	client, err := search.NewClient(appID, apiKey)
	if err != nil {
		// The client can fail to initialize if you pass an invalid parameter.
		fmt.Printf("Client error")
		panic(err)
	}

	// Exporting the rules 
	// https://www.algolia.com/doc/api-reference/api-methods/export-rules/#examples
	fmt.Printf("Original rules:\n")

	var rules []search.Rule

	err = client.BrowseRules(
		indexName,
		*search.NewEmptySearchRulesParams(),
		search.WithAggregator(func(r any, err error) {
			if err != nil {
				log.Fatalf("There was an error: %v", err)
			}

			rules = append(rules, r.(*search.SearchRulesResponse).Hits...)
		}),
	)
	if err != nil {
		log.Fatal(err)
	}

	for _, rule := range rules {
		fmt.Printf("- Rule: %s\n", rule.ObjectID)
	}

	// Adding a new rule 
	// https://www.algolia.com/doc/api-reference/api-methods/save-rule/#save-a-rule
	objectID := "id1"
	fmt.Println("Adding a new rule called:", objectID)
	resSaveRule, err := client.SaveRule(client.NewApiSaveRuleRequest(
	indexName, objectID,
	search.NewEmptyRule().SetObjectID(objectID).SetConditions(
		[]search.Condition{*search.NewEmptyCondition().SetPattern("flower").SetAnchoring(search.Anchoring("contains"))}).SetConsequence(
		search.NewEmptyConsequence().SetPromote(
		[]search.Promote{*search.PromoteObjectIDAsPromote(
			search.NewEmptyPromoteObjectID().SetObjectID("439957720").SetPosition(0))}))))

	if err != nil {
		PrintErrAndExit(err)
	}
	_, err = client.WaitForTask(indexName, resSaveRule.TaskID)
	if err != nil {
		PrintErrAndExit(err)
	}


	// print the response
	fmt.Println(resSaveRule)

	// Exporting the modified rules 
	// https://www.algolia.com/doc/api-reference/api-methods/export-rules/#examples
	fmt.Printf("Updated rules:\n")

	var updatedRules []search.Rule

	err = client.BrowseRules(
		indexName,
		*search.NewEmptySearchRulesParams(),
		search.WithAggregator(func(r any, err error) {
			if err != nil {
				log.Fatalf("There was an error: %v", err)
			}

			updatedRules = append(updatedRules, r.(*search.SearchRulesResponse).Hits...)
		}),
	)
	if err != nil {
		log.Fatal(err)
	}

	for _, rule := range updatedRules {
		fmt.Printf("- Rule: %s\n", rule.ObjectID)
	}
}

File: java/src/main/java/Rules.java
import java.io.IOException;
import java.util.Arrays;
import java.util.concurrent.ExecutionException;

import com.algolia.api.SearchClient;
import com.algolia.model.search.*;

import io.github.cdimascio.dotenv.Dotenv;

public class Rules {
    
    public static void main(String[] args) throws ExecutionException, InterruptedException, IOException {
        Rules.run();
    }

    public static void run() throws ExecutionException, InterruptedException, IOException {
    
        Dotenv dotenv = Dotenv.configure().load();

        String appID = dotenv.get("ALGOLIA_APP_ID");
        String apiKey = dotenv.get("ALGOLIA_API_KEY");
        String indexName = dotenv.get("ALGOLIA_INDEX_NAME");

        // Start the API client
        // https://www.algolia.com/doc/libraries/sdk/methods/search#java
        try (SearchClient client = new SearchClient(appID, apiKey)) {
            // Exporting the rules 
            // https://www.algolia.com/doc/api-reference/api-methods/export-rules/#examples
            System.out.println("Original rules:\n");
            Iterable<Rule> response = client.browseRules(indexName, new SearchRulesParams());

            for (Rule rule : response) {
                System.out.println("- Rule: " + rule.getObjectID());
            }

            String ruleID = "a-rule-id";

            Rule newRule = new Rule()
                    .setObjectID(ruleID)
                    .setConditions(Arrays.asList(new Condition().setPattern("flower").setAnchoring(Anchoring.CONTAINS)))
                    .setConsequence(new Consequence().setPromote(Arrays.asList(new PromoteObjectID().setObjectID("439957720").setPosition(0))));

            // Adding a new rule 
            // https://www.algolia.com/doc/api-reference/api-methods/save-rule/#save-a-rule
            UpdatedAtResponse saveRuleResp = client.saveRule(
                indexName,
                ruleID,
                newRule
            );

            client.waitForTask(indexName, saveRuleResp.getTaskID());

            // Exporting the modified rules 
            // https://www.algolia.com/doc/api-reference/api-methods/export-rules/#examples
            System.out.println("Updated rules:\n");
            response = client.browseRules(indexName, new SearchRulesParams());

            for (Rule rule : response) {
                System.out.println("- Rule: " + rule.getObjectID());
            }

            client.close();
        }
    }
}
